gcp-pubsub/src/test/groovy/io/micronaut/gcp/pubsub/bind/PubSubPushSpec.groovy

```diff
@@ -0,0 +1,271 @@
+    @Client("/")
+    HttpClient pushClient
+
+    @Inject
+    PushConsumer consumer
+
+    void setup() {
+        consumer.msg = null
+    }
+
+    void "a simple push message can be received"() {
+        given:
+        String data = JsonMapper.createDefault().writeValueAsString(Map.of("key", "success"))
+        String encodedData = Base64.getEncoder().encodeToString(data.getBytes())
+        PushRequest request = new PushRequest("projects/test-project/subscriptions/foo", new PushRequest.PushMessage(Map.of("foo", "bar"), encodedData, "1", "2021-02-26T19:13:55.749Z"))
+
+        when:
+        HttpResponse response = pushClient.toBlocking().exchange(HttpRequest.POST("/push", request))
+
+        then:
+        response.status() == HttpStatus.OK
+        consumer.msg
+
+        when:
+        Map<String, String> received = consumer.msg as Map<String, String>
+
+        then:
+        received.get("key") == "success"
+    }
+
+    void "a simple push message with no attributes but non-empty data can be received"() {
+        given:
+        String data = JsonMapper.createDefault().writeValueAsString(Map.of("key", "success"))
+        String encodedData = Base64.getEncoder().encodeToString(data.getBytes())
+        PushRequest request = new PushRequest("projects/test-project/subscriptions/foo", new PushRequest.PushMessage(new HashMap<>(), encodedData, "1", "2021-02-26T19:13:55.749Z"))
+
+        when:
+        HttpResponse response = pushClient.toBlocking().exchange(HttpRequest.POST("/push", request))
+
+        then:
+        response.status() == HttpStatus.OK
+        consumer.msg
+
+        when:
+        Map<String, String> received = consumer.msg as Map<String, String>
+
+        then:
+        received.get("key") == "success"
+    }
+
+    void "a simple push message with no data but non-empty attributes can be received"() {
+        given:
+        PushRequest request = new PushRequest("projects/test-project/subscriptions/bar", new PushRequest.PushMessage(Map.of("foo", "bar"), null, "1", "2021-02-26T19:13:55.749Z"))
+
+        when:
+        HttpResponse response = pushClient.toBlocking().exchange(HttpRequest.POST("/push", request))
+
+        then:
+        response.status() == HttpStatus.OK
+        consumer.msg == "bar"
+    }
+
+    void "an invalid push message is rejected"() {
+        given:
+        String jsonMessage = JsonMapper.createDefault().writeValueAsString([
+                "subscription" : "projects/test-project/subscriptions/foo",
+                "message" : [
+                        "attributes" : new HashMap<>(),
+                        "data" : "",
+                        "messageId" : "1",
+                        "publishTime" : "2021-02-26T19:13:55.749Z"
+                ]
+        ])
+
+        when:
+        pushClient.toBlocking().exchange(HttpRequest.POST("/push", jsonMessage))
+
+        then:
+        HttpClientResponseException ex = thrown()
+        ex.status == HttpStatus.BAD_REQUEST
+    }
```

Comment:
Is it worth testing the response message, to ensure it's a validation error?

```suggestion
        ex.response.getBody(Map).get()._embedded.errors[0].message.contains('message must contain either a non-empty data field or at least one attribute')
    }
```

The message starts with `message.message.null` which feels wrong, but that's not an issue with this PR

Reply:
I attempted to do that at one point, but was having a hard time getting to the message on the response side. It doesn't strictly matter, as the PubSub service blindly treats any non-20x status as a `nack()`. `HttpStatus.BAD_REQUEST` on its own distinguishes this scenario from "normal" `nack()` signals that happen when the payload is valid, as I'm returning `HttpStatus.UNPROCESSABLE_ENTITY` for those.

That said, I might make another attempt for the sake of thoroughness.

Reply:
I'm still unable to verify those validation messages in my tests, and not sure why...when I debug on the server, I can see them getting added to the request. Am I missing a dependency or something that makes those messages available in the client? Unless it's an obvious fix, I'm inclined to let it go.

Reply:
Weird...  Just pasting the change above works for me...

Reply:
I must have been doing something daft, because now it works. Thanks for the assist.